Fixed #33491 -- Fixed change-list selected row-highlight on cancelled delete.

mgaligniana opened the pull request:
[Ticket](https://code.djangoproject.com/ticket/33491)

First, thank you Mariusz for **Cc**me in this ticket!

~~In this PR I apply your suggestion of~~
> ~~we should enforce clearing checkboxes, instead of highlighting rows on Chrome. This would make this behavior browser-agnostic.~~

To run the test please do: `./runtests.py admin_views.tests.SeleniumTests.test_checkboxes_state_after_submit_and_go_back --selenium=<browser> --parallel=1` with `browser` as `chrome`, `firefox` or `safari`

mgaligniana commented on django/contrib/admin/static/admin/js/actions.js:
```diff
@@ -179,23 +179,16 @@
-    ready(function() {
+    };
+
+    window.onpageshow = function() {
```
I replaced the `ready` or the `load` functions by `onpageshow` because Safari doesn't trigger any of them after do a browser back

mgaligniana commented on django/contrib/admin/helpers.py:
```diff
@@ -36,7 +36,9 @@ class ActionForm(forms.Form):
 
-checkbox = forms.CheckboxInput({"class": "action-select"}, lambda value: False)
+checkbox = forms.CheckboxInput(
+    {"class": "action-select", "autocomplete": "off"}, lambda value: False
```
I made use of https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#autocomplete

mgaligniana commented on django/contrib/admin/options.py:
```diff
@@ -962,7 +962,11 @@ def log_deletion(self, request, obj, object_repr):
-    @display(description=mark_safe('<input type="checkbox" id="action-toggle">'))
+    @display(
+        description=mark_safe(
+            '<input type="checkbox" id="action-toggle" autocomplete="off">'
```
I made use of https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#autocomplete

mgaligniana commented:
buildbot, test on selenium.

carltongibson reviewed (commented):
Hi @mgaligniana — thanks for this. The patch looks small, so ... provisionally 👍 

Erm... 😬

But I kind of think we should `wontfix` this. 

Selecting rows from the change list can be tricky. I select many. I go to the confirmation page and realise I've included one extra row by accident. I click to go back. 

Currently, my work is preserved (at least on Safari and Edge testing now, but perhaps not Firefox, consulting with @felixxm). I uncheck the single incorrect row and proceed. 

With this patch I'd need to reselect all the rows again, which would be equally liable to error I'd guess. 

This may be browser-specific, but I'm imagining reports if we change the behaviour here. 

So overall I'd probably be -0 here. (But not -1 if you all feel different) 🤔

The initial report concerned the highlighting colour, which isn't present on the return to the change list. This is added via the `tr.selected` class. I wonder if a bit of [CSS `:has()`](https://caniuse.com/css-has) as an additional selector for that rule would make this nicer for those browsers that support it? 🤔

I'm a bit _Meh_ on the highlight in this case otherwise: I'm not convinced it would be worth lots of JS, for example. 

I don't think we need to enforce around small browser differences here: they're different, we like diversity, and folks get used to the _native_ behaviour of the browser they use. 😜

I hope that's helpful. (As I say 😬)

carltongibson commented:
https://github.com/django/django/blob/35911078fa40eb35859832987fedada76963c01e/django/contrib/admin/static/admin/css/changelists.css#L261-L263

For a `:has()` experiment.

carltongibson commented:
@mgaligniana Just on the `autocomplete` attribute, what's the intent there? 

[From the MDN docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#autocomplete)

> ... This attribute has no effect on input types that do not return numeric or text data, being valid for all input types except checkbox, radio, file, or any of the button types.... 

So this isn't going to do anything in this case is it? 

🤔

mgaligniana commented:
> So this isn't going to do anything in this case is it?
> 
> 🤔

Ohh! I missed that part! But for some reason, in Google and Safari seems it was working because it unchecked the `action-toggle` checkbox, anyway, I updated and now I do it via JS, to be more consitent and have less diff! Thanks Carlton!

Regarding your 1st comment, I agree, the  _"With this patch I'd need to reselect all the rows again"_ could be very annoying for the user experience and it's how is currently working in Firefox for that I also liked the '_agnostic_' idea, but the "_they're different, we like diversity, and folks get used to the native behaviour of the browser they use. 😜_" it's also true! 😵 haha I'm in favor what you decide! 🤗

carltongibson reviewed (commented):
Hey @mgaligniana — thanks for the follow-up. 

* I sceptical about the enforced checkbox clearing here. Personally, I prefer the Safari/Edge/Chrome behaviour over the Firefox one. So if that's the _fix_ I'd vote for `wontfix`. 
* However, can we use `:has()`, in the location I gave, to re-add the highlight to selected rows, even without the additional class? — This would solve the original issue from the ticket I think... (_When I go back, the yellow highlight is missing_).
   * If we were able to do this, once all browsers support `:has()` we'd be able to drop the JS adding the `tr.selected` class.

mgaligniana commented:
buildbot, test on selenium.

mgaligniana commented:
@carltongibson I added the `:has()` and updated the tests, but they are too specific to each browser: as each browser starts to support the `has` property should we start deleting the conditionals?

mgaligniana commented on tests/admin_views/tests.py:
```diff
@@ -6471,6 +6471,92 @@ def test_redirect_on_add_view_continue_button(self):
+        from selenium.webdriver.common.by import By
+        from selenium.webdriver.support.ui import Select
+
+        HIGHLIGHTED = ["rgb(255, 255, 204)", "rgba(255, 255, 204, 1)"]
```
Maybe testing the color is not necessary 🤔

carltongibson reviewed (approved):
Hi @mgaligniana — great. Thanks for the follow-up. 

I combined the two selectors, and adjusted the comment slightly. 
I dropped the test: it's more complex than helpful I think. (But good effort! 👍) 

Let's have it! 🎁 


<details>
<summary>Diff for not-added test case...</summary>

```
    def test_checkboxes_state_after_submit_and_go_back(self):
        from selenium.webdriver.common.by import By
        from selenium.webdriver.support.ui import Select

        HIGHLIGHTED = ["rgb(255, 255, 204)", "rgba(255, 255, 204, 1)"]

        Book.objects.create(name="Book 1")
        Book.objects.create(name="Book 2")
        Book.objects.create(name="Book 3")

        self.admin_login(
            username="super",
            password="secret",
            login_url=reverse("admin:admin_views_book_changelist"),
        )

        # Check all Books.
        select_all_checkbox = self.selenium.find_element(By.ID, "action-toggle")
        select_all_checkbox.click()

        # All rows are `selected` (highlighted class) and their checkboxes are checked.
        rows = self.selenium.find_elements(By.CSS_SELECTOR, "#result_list tbody tr")
        self.assertTrue(
            all(
                [
                    row.get_attribute("class") == "selected"
                    and row.value_of_css_property("background-color") in HIGHLIGHTED
                    and row.find_element(By.CSS_SELECTOR, "input").is_selected()
                    for row in rows
                ]
            )
        )

        # Choose 'Delete selected books'.
        actions_dropdown = Select(
            self.selenium.find_element(By.CSS_SELECTOR, '[name="action"]')
        )
        actions_dropdown.select_by_value("delete_selected")

        # Submit and go back.
        form = self.selenium.find_element(By.ID, "changelist-form")
        with self.wait_page_loaded():
            form.submit()
        self.selenium.back()

        rows = self.selenium.find_elements(By.CSS_SELECTOR, "#result_list tbody tr")

        # Safari keep checkboxes state.
        if self.browser == "safari":
            self.assertTrue(
                all(
                    [
                        row.get_attribute("class") == "selected"
                        and row.value_of_css_property("background-color") in HIGHLIGHTED
                        and row.find_element(By.CSS_SELECTOR, "input").is_selected()
                        for row in rows
                    ]
                )
            )
        # Firefox clean all checkboxes state.
        elif self.browser == "firefox":
            self.assertTrue(
                all(
                    [
                        not row.get_attribute("class")
                        and row.value_of_css_property("background-color")
                        not in HIGHLIGHTED
                        and not row.find_element(By.CSS_SELECTOR, "input").is_selected()
                        for row in rows
                    ]
                )
            )
        # Chrome keeps states but doesn't highlight rows.
        elif self.browser == "chrome":
            self.assertTrue(
                all(
                    [
                        not row.get_attribute("class")
                        and row.value_of_css_property("background-color")
                        not in HIGHLIGHTED
                        and row.find_element(By.CSS_SELECTOR, "input").is_selected()
                        for row in rows
                    ]
                )
            )

```

</details>

mgaligniana commented:
@carltongibson 📣 ⚠️ 

If you combine the two selectors, the highlighted doesn't work even when you click the checbox for 1st time! I divided in two different classes because I think as the `has()` is not supported it breaks the whole style rule. To reproduce it you can add my test diff in main and run it (or just open an admin and click any checkbox)

carltongibson commented:
Ah, yes. Good spot @mgaligniana. Thanks. #15965
